src/riak_kv_vnode.erl

```diff
@@ -1476,147 +1478,162 @@ prepare_put(State=#state{vnodeid=VId,
+        {oldobj, OldObj} ->
+            {{false, {OldObj, no_old_object}}, PutArgs, State2};
+        {newobj, NewObj} ->
+            AMObj = enforce_allow_mult(NewObj, BProps),
+            IndexSpecs = get_index_specs(IndexBackend, CacheData, RequiresGet, AMObj, OldObj),
+            ObjToStore0 = maybe_prune_vclock(PruneTime, AMObj, BProps),
+            ObjectToStore = maybe_do_crdt_update(Coord, CRDTOp, ActorId, ObjToStore0),
+            determine_put_result(ObjectToStore, OldObj, Idx, PutArgs, State2, IndexSpecs, IndexBackend)
+    end.
+
+determine_put_result({error, E}, _, Idx, PutArgs, State, _IndexSpecs, _IndexBackend) ->
+    {{fail, Idx, E}, PutArgs, State};
+determine_put_result(ObjToStore, OldObj, _Idx, PutArgs, State, IndexSpecs, IndexBackend) ->
+    {{true, {ObjToStore, OldObj}},
+     PutArgs#putargs{index_specs = IndexSpecs,
+                     is_index    = IndexBackend}, State}.
+
+maybe_prune_vclock(_PruneTime=undefined, RObj, _BProps) ->
+    RObj;
+maybe_prune_vclock(PruneTime, RObj, BProps) ->
+    riak_object:prune_vclock(RObj, PruneTime, BProps).
+
+get_index_specs(_IndexedBackend=true, CacheData, RequiresGet, NewObj, OldObj) ->
+    case CacheData /= undefined andalso
+         RequiresGet == false of
+        true ->
+            NewData = riak_object:index_data(NewObj),
+            riak_object:diff_index_data(NewData,
+                                        CacheData);
+        false ->
+            riak_object:diff_index_specs(NewObj,
+                                         OldObj)
+    end;
+
+get_index_specs(_IndexedBackend=false, _CacheData, _RequiresGet, _NewObj, _OldObj) ->
+    [].
+
+prepare_put_new_object(#state{idx =Idx} = State,
+               #putargs{robj = RObj,
+                        coord=Coord,
+                        starttime=StartTime,
+                        crdt_op=CRDTOp} = PutArgs,
+                       IndexBackend) ->
+    IndexSpecs = case IndexBackend of
+                     true ->
+                         riak_object:index_specs(RObj);
+                     false ->
+                         []
+                 end,
+    {EpochId, State2} = new_key_epoch(State),
+    RObj2 = maybe_update_vclock(Coord, RObj, EpochId, StartTime),
+    RObj3 = maybe_do_crdt_update(Coord, CRDTOp, EpochId, RObj2),
+    determine_put_result(RObj3, no_old_object, Idx, PutArgs, State2, IndexSpecs, IndexBackend).
+
+get_old_object_or_fake(true, Bucket, Key, Mod, ModState, _CacheClock) ->
+    case do_get_object(Bucket, Key, Mod, ModState) of
+        {error, not_found, _UpdModState} ->
+            not_found;
+        {ok, TheOldObj, _UpdModState} ->
+            {ok, TheOldObj}
+    end;
+get_old_object_or_fake(false, Bucket, Key, _Mod, _ModState, CacheClock) ->
+    FakeObj0 = riak_object:new(Bucket, Key, <<>>),
+    FakeObj = riak_object:set_vclock(FakeObj0, CacheClock),
+    {ok, FakeObj}.
+
+determine_requires_get(CacheClock, RObj, IsSearchable) ->
```

Comment:
As with the backend index type, maybe change from returning `boolean()` to returning `requires_get | does_not_require_get`.